File: frontend/src/api/finance.ts
import { request, tokenStore } from './auth'

export interface Account {
  id: string
  account_name: string
  account_type: string
  institution_name: string | null
  current_balance: string
  created_at: string
}

export interface Category {
  id: string
  name: string
  category_type: string
  color: string | null
}

export interface Tx {
  id: string
  account_id: string
  category_id: string | null
  transaction_type: string
  merchant: string
  description: string | null
  amount: string
  transaction_date: string
}

function authed() {
  return { Authorization: `Bearer ${tokenStore.get()}` }
}

export const financeApi = {
  accounts: () => request<Account[]>('/accounts', { headers: authed() }),
  createAccount: (data: {
    account_name: string
    account_type: string
    current_balance: number
    credit_limit?: number
  }) =>
    request<Account>('/accounts', {
      method: 'POST',
      headers: authed(),
      body: JSON.stringify(data),
    }),
  categories: () => request<Category[]>('/categories', { headers: authed() }),
  transactions: () => request<Tx[]>('/transactions', { headers: authed() }),
  createTransaction: (data: {
    account_id: string
    category_id: string | null
    transaction_type: string
    merchant: string
    description: string | null
    amount: number
    transaction_date: string
  }) =>
    request<Tx>('/transactions', {
      method: 'POST',
      headers: authed(),
      body: JSON.stringify(data),
    }),
  creditProfiles: () =>
    request<CreditProfile[]>('/credit/profiles', { headers: authed() }),
  createCreditProfile: (accountId: string, creditLimit: number) =>
    request<CreditProfile>('/credit/profile', {
      method: 'POST',
      headers: authed(),
      body: JSON.stringify({ account_id: accountId, credit_limit: creditLimit }),
    }),
}

export interface CreditProfile {
  id: string
  account_id: string
  account_name: string
  credit_limit: string
  current_balance: string
  utilization: string
}

File: backend/app/schemas/finance.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class AccountCreate(BaseModel):
    account_name: str = Field(min_length=1, max_length=100)
    account_type: str = Field(min_length=1, max_length=50)
    institution_name: Optional[str] = None
    current_balance: Decimal = Decimal("0")
    credit_limit: Optional[Decimal] = Field(default=None, gt=0)


class CreditProfileOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    account_id: uuid.UUID
    account_name: str
    credit_limit: Decimal
    current_balance: Decimal
    utilization: Decimal


class CreditProfileCreate(BaseModel):
    account_id: uuid.UUID
    credit_limit: Decimal = Field(gt=0)


class AccountOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    account_name: str
    account_type: str
    institution_name: Optional[str]
    current_balance: Decimal
    created_at: datetime


class CategoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    category_type: str
    color: Optional[str]


class TransactionCreate(BaseModel):
    account_id: uuid.UUID
    category_id: Optional[uuid.UUID] = None
    transaction_type: str = "expense"
    merchant: str = Field(min_length=1, max_length=150)
    description: Optional[str] = None
    amount: Decimal = Field(gt=0)
    transaction_date: date


class TransactionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    account_id: uuid.UUID
    category_id: Optional[uuid.UUID]
    transaction_type: str
    merchant: Optional[str]
    description: Optional[str]
    amount: Decimal
    transaction_date: date
